File: solutions/scala/robot-simulator/1/src/main/scala/RobotSimulator.scala
case class Robot(
  val bearing: Bearing,
  val coordinates: (Int, Int),
) {
  def turnLeft: Robot = {
    val updatedBearing = bearing match {
      case Bearing.South => Bearing.East
      case Bearing.East => Bearing.North
      case Bearing.North => Bearing.West
      case Bearing.West => Bearing.South
    }
    this.copy(bearing = updatedBearing)
  }

  def turnRight: Robot = 
    turnLeft.turnLeft.turnLeft

  def advance: Robot = {
    val (x, y) = coordinates
    val updatedCoordinates = bearing match {
      case Bearing.South => (x, y - 1)
      case Bearing.East => (x + 1, y)
      case Bearing.North => (x, y + 1)
      case Bearing.West => (x - 1, y)
    }
    this.copy(coordinates = updatedCoordinates)
  }

  def simulate(commands: String) =
    commands.foldLeft(this) { case (acc, command) =>
      command match {
        case 'R' => acc.turnRight
        case 'L' => acc.turnLeft
        case 'A' => acc.advance
        case _ => throw new IllegalArgumentException("")
      }
    }
}

enum Bearing {
  case South, North, West, East
}

File: solutions/python/robot-simulator/1/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
EAST = 0
NORTH = 1
WEST = 2
SOUTH = 3


class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def advance(self):
        x, y = self.coordinates
        if self.direction == EAST:
            self.coordinates = (x + 1, y)
        elif self.direction == NORTH:
            self.coordinates = (x, y + 1)
        elif self.direction == SOUTH:
            self.coordinates = (x, y - 1)
        else:
            self.coordinates = (x - 1, y)

    def move(self, command):
        for c in command:
            match c:
                case "R":
                    self.direction = (self.direction - 1) % 4
                case "L":
                    self.direction = (self.direction + 1) % 4
                case "A":
                    self.advance()
